### PR TITLE
Add E2E test scripts for qwen3-30b model

### PR DESCRIPTION
**Pull request by YixuanWang-99**
# Description

End-to-End test scripts for Qwen3-30b model. 

# Tests
Test with `RUN_ID=2026-06-15-22-21-28`

to_mt.sh, to_hf.sh, pre-training and sft training are tested on single host TPU v6e-8 vm. 

```
# Export your Hugging Face token (if not already set)
export HF_TOKEN=<your Hugging Face access token>

# Create a unique run ID for this batch of tests
export RUN_ID=$(date +%Y-%m-%d-%H-%M-%S)

# Convert the HF checkpoint to MaxText format
bash tests/end_to_end/tpu/qwen3/30b/test_qwen3_to_mt.sh $RUN_ID

# Run the pre-train test
bash tests/end_to_end/tpu/qwen3/30b/test_qwen3.sh $RUN_ID

# Run the SFT test
bash tests/end_to_end/tpu/qwen3/30b/test_qwen3_sft.sh $RUN_ID


# convert MaxText checkpoint back to HF format
bash tests/end_to_end/tpu/qwen3/30b/test_qwen3_to_hf.sh $RUN_ID $CHECKPOINT_PATH
```

The RL test ran on the v5p-128 cluster and completed 5 training steps. However, the full script did not finish because GKE terminated the jobset due to high cluster demand. The xpk command is:

```
#!/bin/bash

# This script launches a Reinforcement Learning (RL) training workload for the
# Qwen3-30B-A3B model on a GKE cluster using XPK.

set -ex

# --- Environment Setup ---
if ! pip show xpk &> /dev/null; then
    echo "xpk not found in the environment. Please install it by running:"
    echo "uv pip install -e .[runner] --resolution=lowest"
    exit 1
fi

export RUN_ID=${1:-$(date +%Y-%m-%d-%H-%M-%S)}
export PROJECT_ID="${PROJECT_ID:-cloud-tpu-multipod-dev}" # GCP project ID where the cluster is deployed
export CLUSTER_NAME="${CLUSTER_NAME:-v5p-64-bodaborg-europe-west4-b}" # Name of your cluster
export ZONE="${ZONE:-europe-west4}" # Zone where your cluster is deployed
export BASE_OUTPUT_DIRECTORY="${BASE_OUTPUT_DIRECTORY:-gs://runner-maxtext-logs/qwen3-30b-a3b-base}"
export BASE_DOCKER_IMAGE="${BASE_DOCKER_IMAGE:-gcr.io/tpu-prod-env-multipod/maxtext_post_training_nightly:latest}" 
export MAXTEXT_CKPT_PATH="${MAXTEXT_CKPT_PATH:-${BASE_OUTPUT_DIRECTORY}/to_maxtext/unscanned/${RUN_ID}/0/items}"
export TPU_TYPE="v5p-64"
export WORKLOAD_NAME="rl-${RUN_ID}"
export MODEL_NAME="${MODEL_NAME:-qwen3-30b-a3b-base}"
export use_pathways="${use_pathways:-True}"
export run_id="${RUN_ID}"

# XLA Flags
XLA_FLAGS="--xla_tpu_dvfs_p_state=7 \
--xla_tpu_scoped_vmem_limit_kib=65536 \
--xla_tpu_num_sparse_cores_for_gather_offloading=1 \
--xla_tpu_bf16_emission_mode=NATIVE_EMISSION \
--xla_tpu_enable_sparse_core_reduce_scatter_v2=true \
--xla_tpu_enable_sparse_core_collective_offload_all_gather=true \
--xla_tpu_enable_sparse_core_collective_offload_2d_all_gather=true \
--xla_tpu_use_tc_device_shape_on_sc=True \
--xla_sc_disable_megacore_partitioning=True \
--xla_tpu_enable_async_collective_fusion_fuse_all_gather=false \
--xla_enable_async_all_gather=true \
--xla_tpu_prefer_async_allgather_to_allreduce=true \
--xla_tpu_enable_sparse_core_collective_offload_all_reduce=true \
--xla_tpu_enable_sparse_core_collective_offload_reduce_scatter=true \
--xla_tpu_enable_sparse_core_collective_offload_3d_all_gather=true \
--xla_tpu_use_single_sparse_core_for_all_gather_offload=true \
--xla_tpu_enable_concurrent_sparse_core_offloading=true \
--xla_tpu_enable_offloading_gather_to_sparsecore=true \
--xla_tpu_sparse_core_all_gather_latency_multiplier=1 \
--xla_tpu_sparse_core_reduce_scatter_latency_multiplier=3 \
--xla_tpu_enable_sparse_core_collective_aggregator=true \
--xla_tpu_enable_latency_hiding_layer_scheduler=true \
--xla_tpu_scheduler_percent_shared_memory_limit=150 \
--xla_tpu_enable_layer_scheduler_for_dependent_collectives=true \
--xla_tpu_enable_sparse_core_collective_offload_nd_reduce_scatter=true \
--xla_tpu_pcie_bandwidth_multiplier=0.03 \
--xla_tpu_enable_sparse_core_offload_queuing_in_lhs=true \
--xla_tpu_enable_multi_compute_overlap_in_layer_scheduler=false \
--xla_tpu_enable_3d_reduce_scatter_decomposer=false"

# Use base64 encoding to prevent any YAML formatting errors with xpk
MAXTEXT_COMMAND_B64=$(cat "$(dirname "$0")/test_qwen3_rl.sh" | base64 -w 0)

# Workload Creation
xpk workload create-pathways \
  --cluster=$CLUSTER_NAME \
  --project=$PROJECT_ID \
  --zone=$ZONE \
  --priority=medium \
  --max-restarts=0 \
  --tpu-type=$TPU_TYPE \
  --num-slices=1 \
  --base-docker-image="${BASE_DOCKER_IMAGE}" \
  --workload="${WORKLOAD_NAME}" \
  --custom-pathways-proxy-server-args='${XLA_FLAGS}' \
  --command="echo ${MAXTEXT_COMMAND_B64} | base64 -d | bash -s -- ${RUN_ID} ${use_pathways}"

```

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: tests/end_to_end/tpu/qwen3/30b/test_qwen3.sh [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+# bash test_qwen3_to_mt.sh $RUN_ID
+# bash test_qwen3.sh $RUN_ID
+
+
+set -ex
+
+run_id=${1:-$(date +%Y-%m-%d-%H-%M-%S)}
+MODEL_NAME='qwen3-30b-a3b-base'
+
+USE_MULTIMODAL=false
```
Comment (SurbhiJainUSC) [inline tests/end_to_end/tpu/qwen3/30b/test_qwen3.sh]: I guess multimodal Qwen3 is not supported. Let's clean up multimodal if from the scripts.

File: tests/end_to_end/tpu/qwen3/30b/test_qwen3_sft.sh [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+# export RUN_ID=$(date +%Y-%m-%d-%H-%M-%S)
+# bash test_qwen3_to_mt.sh $RUN_ID
+# bash test_qwen3_sft.sh $RUN_ID
+
+
+set -ex
+export PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION=python
+export NEW_MODEL_DESIGN=1
+run_id=${1:-$(date +%Y-%m-%d-%H-%M-%S)}
+use_pathways=${2:-false}
```
Comment (SurbhiJainUSC) [inline tests/end_to_end/tpu/qwen3/30b/test_qwen3_sft.sh]: Please remove this variable and hardcode `enable_single_controller=True` as this model won't fit on single VM.

File: tests/end_to_end/tpu/qwen3/30b/test_qwen3_sft.sh [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+MODEL_NAME='qwen3-30b-a3b-base'
+
+BASE_OUTPUT_DIRECTORY=gs://runner-maxtext-logs/${MODEL_NAME}
+UNSCANNED_CKPT_PATH=${BASE_OUTPUT_DIRECTORY}/to_maxtext/unscanned/${run_id}/0/items
+SCANNED_CKPT_PATH=${BASE_OUTPUT_DIRECTORY}/to_maxtext/scanned/${run_id}/0/items
+
+python3 -m maxtext.inference.vllm_decode \
```
Comment (SurbhiJainUSC) [inline tests/end_to_end/tpu/qwen3/30b/test_qwen3_sft.sh:29]: What is the `prompt` for this inference command?

File: tests/end_to_end/tpu/qwen3/30b/test_qwen3_sft.sh [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    model_name=${MODEL_NAME} \
+    load_parameters_path=${UNSCANNED_CKPT_PATH} \
+    vllm_hf_overrides='{architectures: ["MaxTextForCausalLM"]}' \
+    hbm_utilization_vllm=0.85 \
+    use_chat_template=True scan_layers=false enable_single_controller=${use_pathways} \
+    ici_tensor_parallelism=4
+
+python3 -m maxtext.trainers.post_train.sft.train_sft \
```
Comment (SurbhiJainUSC) [inline tests/end_to_end/tpu/qwen3/30b/test_qwen3_sft.sh:38]: Please add comments for each Python command.

File: tests/end_to_end/tpu/qwen3/30b/test_qwen3_sft.sh [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    "${MAXTEXT_CONFIGS_DIR:-${MAXTEXT_REPO_ROOT:-$PWD}/src/maxtext/configs}"/post_train/sft.yml \
```
Comment (SurbhiJainUSC) [inline tests/end_to_end/tpu/qwen3/30b/test_qwen3_sft.sh]: No need to set `"${MAXTEXT_CONFIGS_DIR:-${MAXTEXT_REPO_ROOT:-$PWD}/src/maxtext/configs}"/post_train/sft.yml`

File: tests/end_to_end/tpu/qwen3/30b/test_qwen3_sft.sh [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    base_output_directory=${BASE_OUTPUT_DIRECTORY}/sft \
+    load_parameters_path=${UNSCANNED_CKPT_PATH} \
```
Comment (SurbhiJainUSC) [inline tests/end_to_end/tpu/qwen3/30b/test_qwen3_sft.sh]: Can we run training on `SCANNED_CKPT_PATH`?

File: tests/end_to_end/tpu/qwen3/30b/test_qwen3_sft.sh [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    per_device_batch_size=0.125 run_name=${run_id} \
+    steps=5 scan_layers=false \
+    model_name=${MODEL_NAME} enable_single_controller=${use_pathways} \
+    checkpoint_storage_use_zarr3=False checkpoint_storage_use_ocdbt=False \
+    remat_policy=full use_tunix_gradient_accumulation=False \
```
Comment (SurbhiJainUSC) [inline tests/end_to_end/tpu/qwen3/30b/test_qwen3_sft.sh]: Why do we need to set ` use_tunix_gradient_accumulation=False`?

File: tests/end_to_end/tpu/qwen3/30b/test_qwen3_sft.sh [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    ici_tensor_parallelism=1 ici_fsdp_parallelism=1 ici_expert_parallelism=8 max_target_length=16 weight_dtype=bfloat16 dtype=bfloat16 opt_type=sgd
+
+python3 -m maxtext.inference.vllm_decode \
```
Comment (SurbhiJainUSC) [inline tests/end_to_end/tpu/qwen3/30b/test_qwen3_sft.sh:50]: Please specify `prompt`

File: tests/end_to_end/tpu/qwen3/30b/test_qwen3_rl.sh [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+cat << 'EOF' > usercustomize.py
+import multiprocessing
+try:
+    multiprocessing.set_start_method("spawn", force=True)
+except Exception:
+    pass
+EOF
+export PYTHONPATH=${PYTHONPATH:-.}:$(pwd)
+run_id=${1:-$(date +%Y-%m-%d-%H-%M-%S)}
+use_pathways=${2:-True}
```
Comment (SurbhiJainUSC) [inline tests/end_to_end/tpu/qwen3/30b/test_qwen3_rl.sh]: Same comment as test_qwen3_sft.sh`

File: tests/end_to_end/tpu/qwen3/30b/test_qwen3_rl.sh [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+    vllm_hf_overrides='{architectures: ["MaxTextForCausalLM"]}' \
+    hbm_utilization_vllm=0.85 \
+    prompt='Suggest some famous landmarks in London.' \
+    max_target_length=256 max_num_batched_tokens=256 \
+    ici_tensor_parallelism=8  \
+    allow_split_physical_axes=True prefuse_moe_weights=True \
+    use_chat_template=True scan_layers=False enable_single_controller=${use_pathways}
+
+python3 -m maxtext.trainers.post_train.rl.train_rl \
+    "${MAXTEXT_CONFIGS_DIR:-${MAXTEXT_REPO_ROOT:-$PWD}/src/maxtext/configs}"/post_train/rl.yml \
```
Comment (SurbhiJainUSC) [inline tests/end_to_end/tpu/qwen3/30b/test_qwen3_rl.sh]: No need to set config

File: tests/end_to_end/tpu/qwen3/30b/test_qwen3_rl.sh [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+    base_output_directory=${BASE_OUTPUT_DIRECTORY}/rl \
+    load_parameters_path=${UNSCANNED_CKPT_PATH} \
```
Comment (SurbhiJainUSC) [inline tests/end_to_end/tpu/qwen3/30b/test_qwen3_rl.sh]: Run training on scanned ckpt

File: tests/end_to_end/tpu/qwen3/30b/test_qwen3_rl.sh (hunk 9 above)
Comment (YixuanWang-99) [reply tests/end_to_end/tpu/qwen3/30b/test_qwen3_rl.sh]: Done

File: tests/end_to_end/tpu/qwen3/30b/test_qwen3_rl.sh (hunk 10 above)
Comment (YixuanWang-99) [reply tests/end_to_end/tpu/qwen3/30b/test_qwen3_rl.sh]: removed

File: tests/end_to_end/tpu/qwen3/30b/test_qwen3_rl.sh (hunk 11 above)
Comment (YixuanWang-99) [reply tests/end_to_end/tpu/qwen3/30b/test_qwen3_rl.sh]: changed

File: tests/end_to_end/tpu/qwen3/30b/test_qwen3_sft.sh (hunk 2 above)
Comment (YixuanWang-99) [reply tests/end_to_end/tpu/qwen3/30b/test_qwen3_sft.sh]: Done

File: tests/end_to_end/tpu/qwen3/30b/test_qwen3_sft.sh (hunk 3 above)
Comment (YixuanWang-99) [reply tests/end_to_end/tpu/qwen3/30b/test_qwen3_sft.sh:29]: Added

File: tests/end_to_end/tpu/qwen3/30b/test_qwen3_sft.sh (hunk 4 above)
Comment (YixuanWang-99) [reply tests/end_to_end/tpu/qwen3/30b/test_qwen3_sft.sh:38]: added

File: tests/end_to_end/tpu/qwen3/30b/test_qwen3_sft.sh (hunk 5 above)
Comment (YixuanWang-99) [reply tests/end_to_end/tpu/qwen3/30b/test_qwen3_sft.sh]: removed

File: tests/end_to_end/tpu/qwen3/30b/test_qwen3_sft.sh (hunk 6 above)
Comment (YixuanWang-99) [reply tests/end_to_end/tpu/qwen3/30b/test_qwen3_sft.sh]: Done, fixed all training with scanned ckpt

File: tests/end_to_end/tpu/qwen3/30b/test_qwen3_sft.sh (hunk 7 above)
Comment (YixuanWang-99) [reply tests/end_to_end/tpu/qwen3/30b/test_qwen3_sft.sh]: Not necessary, deleted

File: tests/end_to_end/tpu/qwen3/30b/test_qwen3_sft.sh (hunk 8 above)
Comment (YixuanWang-99) [reply tests/end_to_end/tpu/qwen3/30b/test_qwen3_sft.sh:50]: Done

**Review by SurbhiJainUSC — APPROVED** [review]

**Review by hengtaoguo — APPROVED** [review]

File: tests/end_to_end/tpu/qwen3/30b/test_qwen3_sft.sh [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+# Step 2: Run SFT starting from the pre-converted checkpoint
+python3 -m maxtext.trainers.post_train.sft.train_sft \
+    base_output_directory=${BASE_OUTPUT_DIRECTORY}/sft \
+    load_parameters_path=${SCANNED_CKPT_PATH} \
+    per_device_batch_size=0.125 run_name=${run_id} \
+    steps=5 scan_layers=true \
+    model_name=${MODEL_NAME} \
+    checkpoint_storage_use_zarr3=False checkpoint_storage_use_ocdbt=False \
+    remat_policy=full \
+    ici_tensor_parallelism=1 ici_fsdp_parallelism=1 ici_expert_parallelism=8 \
```
Comment (hengtaoguo) [inline tests/end_to_end/tpu/qwen3/30b/test_qwen3_sft.sh:46]: Which device/cluster is this script targeting? Should we call out in the header if we specify shardings here?

**Review by hengtaoguo — APPROVED** [review]

File: tests/end_to_end/tpu/qwen3/30b/test_qwen3_sft.sh (hunk 12 above)
Comment (YixuanWang-99) [reply tests/end_to_end/tpu/qwen3/30b/test_qwen3_sft.sh:46]: I think the target cluster is v5p-128, I am unsure if we have sort of 'automatic sharding setup' in the backend.
